src/field_theory/complete_ordered.lean

```diff
@@ -0,0 +1,898 @@
+-/
+
+noncomputable theory
+open_locale classical
+
+open set
+
+/-- The natural equiv between the rationals and the rationals as a set inside any characteristic 0
+    division ring -/
+def rat_equiv_of_char_zero (F : Type*) [division_ring F] [char_zero F] : ℚ ≃ range (coe : ℚ → F) :=
+equiv.set.range coe rat.cast_injective
+
+/-- The natural equiv between the rationals as a set inside any pair of characteristic 0
+    division rings -/
+def range_rat_equiv (F K : Type*) [division_ring F] [char_zero F] [division_ring K] [char_zero K] :
+  range (coe : ℚ → F) ≃ range (coe : ℚ → K) :=
+(rat_equiv_of_char_zero F).symm.trans (rat_equiv_of_char_zero K)
+
+@[simp]
+lemma range_rat_equiv_apply {F K : Type*} {q : ℚ}
+  [division_ring F] [char_zero F] [division_ring K] [char_zero K] :
+  range_rat_equiv F K ⟨q, mem_range_self q⟩ = ⟨q, mem_range_self q⟩ :=
+by simp only [range_rat_equiv, rat_equiv_of_char_zero, equiv.set.range_apply,
+  function.comp_app, subtype.mk_eq_mk, equiv.coe_trans, rat.cast_inj, equiv.symm_apply_eq]
+
+/-- The function sending subsets of the rationals embedded inside of one characteristic zero
+    division ring to the corresponding subset of a second such ring. -/
+def cut_map (F K : Type*) [division_ring F] [char_zero F] [division_ring K] [char_zero K] :
+  set (set.range (coe : ℚ → F)) → set (set.range (coe : ℚ → K)) :=
+λ c, (range_rat_equiv F K) '' c
+
+/- No longer used -/
+lemma bdd_above_add {F : Type*} [ordered_add_comm_monoid F] {A B : set F} :
+  bdd_above A → bdd_above B → bdd_above (A + B) :=
+begin
+  rintros ⟨bA, hbA⟩ ⟨bB, hbB⟩,
+  use bA + bB,
+  rintros x ⟨xa, xb, hxa, hxb, rfl⟩,
+  exact add_le_add (hbA hxa) (hbB hxb),
+end
+
+/-- Homomorphism commuting with multiplicative, additive and order structure. -/
+@[nolint has_inhabited_instance]
+structure ordered_ring_hom (R S : Type*) [ordered_semiring R] [ordered_semiring S]
+extends R →+* S :=
+(map_rel' : ∀ {a b}, a ≤ b → to_fun a ≤ to_fun b)
+
+/-- Reinterpret an ordered ring homomorphism as a ring homomorphism. -/
+add_decl_doc ordered_ring_hom.to_ring_hom
+
+namespace ordered_ring_hom
+
+infix ` →+*o `:25 := ordered_ring_hom
+
+variables {R S : Type*} [ordered_semiring R] [ordered_semiring S]
+
+/-- Reinterpret an ordered ring homomorphism as an order homomorphism. -/
+def to_rel_hom (f : R →+*o S) : ((≤) : R → R → Prop) →r ((≤) : S → S → Prop) := { ..f }
+
+instance has_coe_to_ring_hom : has_coe (R →+*o S) (R →+* S) := ⟨to_ring_hom⟩
+instance has_coe_to_hom : has_coe (R →+*o S) (R → S) := ⟨ring_hom.to_fun ∘ to_ring_hom⟩
+instance has_coe_to_rel_hom : has_coe (R →+*o S) (((≤) : R → R → Prop) →r ((≤) : S → S → Prop)) :=
+⟨to_rel_hom⟩
+
+instance : has_coe_to_fun (R →+*o S) := ⟨_, λ f, f.to_fun⟩
+
+@[simp] lemma to_ring_hom_eq_coe {f : R →+*o S} : f.to_ring_hom = f := rfl
+@[simp] lemma to_order_iso_eq_coe {f : R →+*o S} : f.to_rel_hom = f := rfl
+@[simp] lemma coe_ring_hom_to_fun_eq_coe_fun {f : R →+*o S} : (f : R →+* S).to_fun = f := rfl
```

Comment:
Shouldn't the linter complain about this one? `ring_hom.to_fun` shouldn't be simp normal, and if it is then this lemma should be just about ring homes.

Reply:
I guess this lemma just applies before the other one fires right?

Reply:
I don't know if that's a problem, the simp nf linter is meant to catch lemmas that will never fire but this one does if it sees a term like the lhs here?

Reply:
I was expecting [ring_hom.to_fun_eq_coe](https://leanprover-community.github.io/mathlib_docs/algebra/ring/basic.html#ring_hom.to_fun_eq_coe) to fire here

Reply:
Yeah I'm not sure why this takes priority over that, I think they have the same numerical priority, 1000